### PR TITLE
[CLIPPER-18] Implement and test standard selection policies

### PR DESCRIPTION
**Pull request by yujialuo**

### CODE REVIEW

File: src/frontends/src/query_frontend.hpp [hunk 1]
```diff
@@ -84,17 +85,18 @@ std::shared_ptr<Input> decode_input(InputType input_type, ptree& parsed_json) {
 }
 
 Output decode_output(OutputType output_type, ptree parsed_json) {
```
Comment (dcrankshaw) [inline src/frontends/src/query_frontend.hpp]: Delete this function

File: src/frontends/src/query_frontend.hpp [hunk 2]
```diff
@@ -262,7 +264,7 @@ class RequestHandler {
     std::shared_ptr<Input> input = decode_input(input_type, pt);
```
Comment (dcrankshaw) [inline src/frontends/src/query_frontend.hpp]: Remove `model_name` and `model_version` from the JSON format comment on this method.

File: src/frontends/src/query_frontend.hpp [hunk 3]
```diff
@@ -262,7 +264,7 @@ class RequestHandler {
...
     Output output = decode_output(output_type, pt);
```
Comment (dcrankshaw) [inline src/frontends/src/query_frontend.hpp]: We don't actually need to create an output object, just parse the label:

```
double y_hat = pt.get<double>("label");
``

File: src/frontends/src/query_frontend.hpp [hunk 4]
```diff
@@ -262,7 +264,7 @@ class RequestHandler {
...
     auto update = query_processor_.update(FeedbackQuery{
-        name, uid, {std::make_pair(input, output)}, policy, models});
+        name, uid, {Feedback(input, output.y_hat_)}, policy, models});
```
Comment (dcrankshaw) [inline src/frontends/src/query_frontend.hpp]: Replace `output.y_hat_` with `y_hat`

File: src/libclipper/CMakeLists.txt [hunk 5]
```diff
@@ -42,7 +42,8 @@ add_executable(libclippertests EXCLUDE_FROM_ALL
     test/persistent_state_test.cpp
     test/redis_test.cpp
     test/config_test.cpp
-    test/selection_policies_test.cpp
+    # test/selection_policies_test.cpp
```
Comment (dcrankshaw) [inline src/libclipper/CMakeLists.txt]: Delete the commented out line

File: src/libclipper/include/clipper/datatypes.hpp [hunk 6]
```diff
@@ -238,7 +238,22 @@ class Response {
   std::vector<VersionedModelId> models_used_;
 };
 
-using Feedback = std::pair<std::shared_ptr<Input>, Output>;
+// using Feedback = std::pair<std::shared_ptr<Input>, Output>;
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/datatypes.hpp]: Delete the commented out line

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+#include <map>
+#include <memory>
+#include <unordered_map>
+
+#include "datatypes.hpp"
+#include "task_executor.hpp"
+
+namespace clipper {
+//// State Data Structure
+using ModelInfo = std::unordered_map<std::string, double>;
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: What is `ModelInfo`?

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+  static State deserialize_state(const std::string& bytes) {
+    return Derived::deserialize_state(bytes);
+  }
+
+  /**
+   * Human readable debug string for the state
+   */
+  static std::string state_debug_string(const State& state) {
+    return Derived::state_debug_string(state);
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: Can you implement `state_debug_string` for each implementation? The string should be a human-readable representation of the state.

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+
+  static Exp3State initialize(
+      const std::vector<VersionedModelId>& candidate_models);
+
+  static Exp3State add_models(Exp3State state,
+                              const std::vector<VersionedModelId>& new_models);
+
+  static long hash_models(
+      const std::vector<VersionedModelId>& candidate_models) {
+    UNUSED(candidate_models);
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: remove unused and comment out the candidate_models argument (`/*candidate_models*/`)

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+  static Exp4State initialize(
+      const std::vector<VersionedModelId>& candidate_models);
+
+  static Exp4State add_models(Exp4State state,
+                              const std::vector<VersionedModelId>& new_models);
+
+  static long hash_models(
+      const std::vector<VersionedModelId>& candidate_models) {
+    UNUSED(candidate_models);
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: remove UNUSED

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+  static EpsilonGreedyState initialize(
+      const std::vector<VersionedModelId>& candidate_models);
+
+  static EpsilonGreedyState add_models(
+      EpsilonGreedyState state,
+      const std::vector<VersionedModelId>& new_models);
+
+  static long hash_models(
+      const std::vector<VersionedModelId>& candidate_models) {
+    UNUSED(candidate_models);
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: Remove UNUSED

File: src/libclipper/src/selection_policies.cpp [hunk 12]
```diff
@@ -1,344 +1,407 @@
...
 }
 
-long NewestModelSelectionPolicy::hash_models(
-    const std::vector<VersionedModelId>& candidate_models) {
-  UNUSED(candidate_models);
-  return 0;
+VersionedModelId Exp3Policy::select(Exp3State state,
+                                    std::vector<VersionedModelId>& models) {
+  // Helper function for randomly drawing an arm based on its normalized weight
+  auto rand_num = rand() % 1;
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: What is the `% 1` for?

File: src/libclipper/src/selection_policies.cpp [hunk 13]
```diff
@@ -1,344 +1,407 @@
...
+  VersionedModelId selected_model;
+  auto it = models.begin();
+  while (rand_num >= 0) {
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: Can you add a comment explaining what this while loop is doing?

File: src/libclipper/src/selection_policies.cpp [hunk 14]
```diff
@@ -1,344 +1,407 @@
...
-  if (predictions.empty()) {
-    return Output{0.0, std::make_pair("none", 0)};
-  } else {
-    return predictions.front();
-  }
+Output Exp3Policy::combine_predictions(Exp3State /*state*/, Query /*query*/,
+                                       std::vector<Output> predictions) {
+  // Don't need to do anything
+  // TODO: what if predictions is empty?
+  return predictions.front();
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: What happens if `predictions` is empty?

File: src/libclipper/src/selection_policies.cpp [hunk 15]
```diff
@@ -1,344 +1,407 @@
...
-long NewestModelSelectionPolicy::hash_models(
-    const std::vector<VersionedModelId>& candidate_models) {
-  UNUSED(candidate_models);
-  return 0;
+VersionedModelId Exp3Policy::select(Exp3State state,
+                                    std::vector<VersionedModelId>& models) {
+  // Helper function for randomly drawing an arm based on its normalized weight
+  auto rand_num = rand() % 1;
+  VersionedModelId selected_model;
+  auto it = models.begin();
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: What about the case where `models` is empty? You should log an error message and select prediction tasks should return an empty vector.

File: src/libclipper/src/selection_policies.cpp [hunk 16]
```diff
@@ -1,344 +1,407 @@
...
-}
-
-VersionedModelId NewestModelSelectionPolicy::process_feedback(
-    VersionedModelId state, Feedback feedback,
-    std::vector<Output> predictions) {
-  UNUSED(feedback);
-  UNUSED(predictions);
-  return state;
+Exp3Policy::select_feedback_tasks(Exp3State state, FeedbackQuery feedback,
+                                  long query_id) {
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: Handle the case where there are no candidate models

File: src/libclipper/src/selection_policies.cpp [hunk 17]
```diff
@@ -1,344 +1,407 @@
...
+  // Predict Task
+  auto selected_model = select(state, feedback.candidate_models_);
+  auto task1 =
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: let's not call these variables `task1` and `task2`. How about `predict_tasks` and `feedback_tasks`?

File: src/libclipper/src/selection_policies.cpp [hunk 18]
```diff
@@ -1,344 +1,407 @@
...
+// ********
+// * EXP3 *
+// ********
+// Descripton: Single Model Selection Policy, iteratively update weights
+Exp3State Exp3Policy::initialize(
+    const std::vector<VersionedModelId>& candidate_models_) {
+  // Exp3: sum of weights; each model has "weight", "max loss"
+  Map map(candidate_models_.size(), &versioned_model_hash);
+  for (VersionedModelId id : candidate_models_) {
+    ModelInfo info = {{"weight", 1.0}, {"max loss", 0.0}};
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: can you rename `max loss` to `max_loss`?

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+
+#include "datatypes.hpp"
+#include "task_executor.hpp"
+
+namespace clipper {
+//// State Data Structure
+using ModelInfo = std::unordered_map<std::string, double>;
+using Map = std::unordered_map<VersionedModelId, ModelInfo,
+                               std::function<size_t(const VersionedModelId&)>>;
+// Exp3: sum of weights; each model has "weight", "max loss"
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: Let's make the state objects classes instead of nested maps and pairs. Then we can give all the fields nice names and avoiding referencing things as `first` and `second`.

File: src/libclipper/test/policy_test.cpp [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+
+  // Select
+  auto query = Utility::create_query(models);
+  auto tasks = Exp3Policy::select_predict_tasks(exp3state, query, 1000);
+  ASSERT_NE(model_3.second, tasks.front().model_.second);
+  
+  // Serialization
+  // auto bytes = Exp3Policy::serialize_state(exp3state);
+  // auto new_state = Exp3Policy::deserialize_state(bytes);
+  // ASSERT_EQ(exp3state.first, new_state.first);
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Can you add a serialization test as a separate test?

File: src/libclipper/test/policy_test.cpp [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+}
+
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Add a serialization test, an update test, and a selection test for each selection policy.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
Finished the first round of reviews. Overall it looks good. I added comments inline but the big ones are:

+ Add more tests. We should have a selection test, an update test, and a serialization test for each policy.
+ Let's change the State objects to classes instead of nested pairs and maps (which make the code very difficult to read).
+ Handle edge cases (e.g. when vectors could be empty).

I added a lot of comments to the `Exp3Policy` implementation. Most of those apply to all of the policy implementations.

Finally, can you add a comment to each policy that includes a link to more information about the algorithm (a link to the relevant paper would be great).

File: src/libclipper/src/selection_policies.cpp (hunk 16 above)
Comment (yujialuo) [inline src/libclipper/src/selection_policies.cpp]: i feel like the code won't go wrong even if there are no candidate models

**Comment by dcrankshaw** [comment]
Why did you close this?

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 22]
```diff
@@ -0,0 +1,307 @@
...
+
+namespace clipper {
+
+// *********
+// * State *
+// *********
+
+// Indicate information of a model
+// Key as info name, value as info value
+using ModelInfo = std::unordered_map<std::string, double>;
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: What is the value here? What information about a model does this map store?

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 23]
```diff
@@ -0,0 +1,307 @@
...
+// A map of all models to corresponding model info
+using Map = std::unordered_map<VersionedModelId, ModelInfo,
+std::function<size_t(const VersionedModelId&)>>;
+
+class PolicyState {
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: Can we call this BanditPolicyState? PolicyState makes it sound like all policies need to use this class to store their state and that's not the case.

File: src/libclipper/src/selection_policies.cpp [hunk 24]
```diff
@@ -1,344 +1,484 @@
...
+  return state;
+};
+
+std::string PolicyState::debug_string() const {
 
-VersionedModelId NewestModelSelectionPolicy::add_models(
-    VersionedModelId state, std::vector<VersionedModelId> new_models) {
-  UNUSED(state);
-  return new_models.front();
+  // State string example: "3.0;model1,1.0,0.4;model2,1.0,0.4;......."
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: What are the meanings of the 2 numbers following the model name? Is it (name, version, weight)?

File: src/libclipper/src/selection_policies.cpp [hunk 25]
```diff
@@ -1,344 +1,484 @@
...
+  auto y_hat = 0;
+  std::vector<VersionedModelId> models;
+  for (auto p : predictions) {
+    auto model_id = (p.models_used_).front();
+    y_hat += (state.model_map_[model_id]["weight"] / state.weight_sum_) * p.y_hat_;
+    models.push_back(model_id);
   }
+  auto output = Output(y_hat, models);
+  return output;
 }
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: I don't think this averaging will quite work. You need to round to the nearest label value (e.g. if the labels are -1, +1 and the average is -0.2, you need to round to -1.0). If this implementation only works for binary classification right now that's fine, but you should add a comment noting that in the `selection_policies.hpp`.

File: src/libclipper/src/selection_policies.cpp [hunk 26]
```diff
@@ -1,344 +1,484 @@
...
-  UNUSED(feedback);
-  UNUSED(predictions);
+Exp3Policy::select_feedback_tasks(PolicyState& state, FeedbackQuery feedback,
+                                  long query_id) {
+  // Predict Task
+  auto selected_model = select(state);
+  auto predict_task = PredictTask(feedback.feedback_.input_, selected_model, -1, query_id, -1);
+  std::vector<PredictTask> predict_tasks{predict_task};
+  // Feedback Task
+  auto feedback_task = FeedbackTask(feedback.feedback_, selected_model, query_id, -1);
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: Don't send any feedback tasks for now. The functionality for this API isn't implemented yet.

File: src/libclipper/src/selection_policies.cpp [hunk 27]
```diff
@@ -1,344 +1,484 @@
...
+Exp4Policy::select_feedback_tasks(PolicyState& /*state*/,
+                              FeedbackQuery feedback,
+                              long query_id) {
+  std::vector<PredictTask> predict_tasks;
+  std::vector<FeedbackTask> feedback_tasks;
+  for (VersionedModelId id : feedback.candidate_models_) {
+    auto predict_task = PredictTask(feedback.feedback_.input_, id, -1, query_id, -1);
+    auto feedback_task = FeedbackTask(feedback.feedback_, id, query_id, -1);
+    predict_tasks.push_back(predict_task);
+    feedback_tasks.push_back(feedback_task);
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: Don't send feedback tasks

File: src/libclipper/src/selection_policies.cpp [hunk 28]
```diff
@@ -1,344 +1,484 @@
...
-  ss << (state.end() - 1)->first.first << ITEM_PART_CONCATENATOR
-     << (state.end() - 1)->first.second << ITEM_PART_CONCATENATOR
-     << std::to_string((state.end() - 1)->second);
-  std::cout << "BanditPolicy::serialize_state result: " << ss.str()
-            << std::endl;
-  return ss.str();
+std::vector<PredictTask> UCBPolicy::select_predict_tasks(PolicyState& state,
+                                                         Query query,
+                                                         long query_id) {
+  return Exp3Policy::select_predict_tasks(state, query, query_id);
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: Why are you using `Exp3Policy` to select? Shouldn't you be using `UCBPolicy::select`?

File: src/libclipper/test/policy_test.cpp [hunk 29]
```diff
@@ -0,0 +1,261 @@
...
+            state.model_map_[model_2]["weight"]);
+  ASSERT_GT(state.model_map_[model_2]["weight"],
+            state.model_map_[model_3]["weight"]);
+}
+
+TEST_F(Exp3Test, SelectionTest) {
+  auto query = Utility::create_query(models);
+  auto tasks = Exp3Policy::select_predict_tasks(state, query, 1000);
+  ASSERT_NE(model_3.second, tasks.front().model_.second);
+}
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: What are you testing for here? Can you add more comparisons to check for test success/failure scenarios?

Comment (dcrankshaw) [reply src/libclipper/test/policy_test.cpp]: Instead of just checking that the worst model isn't selected, let's select 100 tasks and check that the most frequently selected model is the best model, the second most frequently selected model is the second best model, etc.

File: src/libclipper/test/policy_test.cpp [hunk 30]
```diff
@@ -0,0 +1,261 @@
...
+  ASSERT_GT(state.model_map_[model_1]["weight"],
+            state.model_map_[model_2]["weight"]);
+  ASSERT_GT(state.model_map_[model_2]["weight"],
+            state.model_map_[model_3]["weight"]);
+}
+
+TEST_F(Exp4Test, SelectionTest) {
+  auto query = Utility::create_query(models);
+  auto tasks = Exp4Policy::select_predict_tasks(state, query, 1000);
+  ASSERT_NE(model_3.second, tasks.front().model_.second);
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Let's do the same thing as for `Exp3Test::SelectionTest`.

File: src/libclipper/test/policy_test.cpp [hunk 31]
```diff
@@ -0,0 +1,261 @@
...
+            state.model_map_[model_2]["expected_loss"]);
+  ASSERT_GT(state.model_map_[model_2]["expected_loss"],
+            state.model_map_[model_1]["expected_loss"]);
+}
+
+TEST_F(EpsilonGreedyTest, SerializationTest) {
+  auto bytes = EpsilonGreedyPolicy::serialize_state(state);
+  auto new_state = EpsilonGreedyPolicy::deserialize_state(bytes);
+  ASSERT_EQ(state.weight_sum_, new_state.weight_sum_);
+}
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: What happened to `EpsilonGreedy::SelectionTest`?

File: src/libclipper/test/policy_test.cpp [hunk 32]
```diff
@@ -0,0 +1,261 @@
...
+  }
+  std::vector<VersionedModelId> models;
+  VersionedModelId model_1 = std::make_pair("classification", 0);
+  VersionedModelId model_2 = std::make_pair("regression", 1);
+  VersionedModelId model_3 = std::make_pair("random_forest", 2);
+  int times = 1000;
+  PolicyState state;
+};
+
+TEST_F(Exp4Test, UpdateTest) {
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: This test fails for me

File: src/libclipper/test/policy_test.cpp [hunk 33]
```diff
@@ -0,0 +1,261 @@
...
+  ASSERT_GT(state.model_map_[model_3]["expected_loss"],
+            state.model_map_[model_2]["expected_loss"]);
+  ASSERT_GT(state.model_map_[model_2]["expected_loss"],
+            state.model_map_[model_1]["expected_loss"]);
+}
+
+TEST_F(UCBTest, SelectionTest) {
+  auto query = Utility::create_query(models);
+  auto tasks = UCBPolicy::select_predict_tasks(state, query, 1000);
+  ASSERT_EQ(model_1.second, tasks.front().model_.second);
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Same comment about selecting 100 tasks as for `Exp3Test::SelectionTest`

File: src/libclipper/test/selection_policies_test.cpp [hunk 34]
```diff
@@ -1,54 +1,54 @@
...
-      state, f3neg, {Output(0.0, good_model), Output(1.0, bad_model)});
-
-  std::cout << BanditPolicy::serialize_state(state);
-
-  ASSERT_EQ(state[0].first, good_model);
-  ASSERT_EQ(state[1].first, bad_model);
-  ASSERT_GT(state[0].second, state[1].second);
-  ASSERT_FLOAT_EQ(state[0].second + state[1].second, 1.0);
-}
+// #include <gtest/gtest.h>
```
Comment (dcrankshaw) [inline src/libclipper/test/selection_policies_test.cpp]: You can delete this file

File: src/management/src/management_frontend.hpp [hunk 35]
```diff
@@ -248,19 +248,22 @@ class RequestHandler {
...
         clipper::redis::str_to_models(app_metadata["candidate_models"]);
     std::string policy = app_metadata["policy"];
-    if (policy == "bandit_policy") {
-      return lookup_selection_state<clipper::BanditPolicy>(
-          state_db_, app_name, uid, candidate_models);
-    } else {
-      return "ERROR: " + app_name +
-             " does not support looking up selection policy state";
-    }
+    return "ERROR: " + app_name +
```
Comment (dcrankshaw) [inline src/management/src/management_frontend.hpp]: Now that we have `debug_string` methods implemented for the selection policies, can you add support for looking them up? Basically just replace the commented out `bandit_policy` if statement branch for each of your policies.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
`Exp4Test.UpdateTest` consistently fails right now. Look into why that's failing.

Other than that, some relatively minor comments and then this should be ready to merge.

File: src/libclipper/src/selection_policies.cpp (hunk 26 above)
Comment (yujialuo) [reply src/libclipper/src/selection_policies.cpp]: Fixed

File: src/libclipper/src/selection_policies.cpp (hunk 28 above)
Comment (yujialuo) [reply src/libclipper/src/selection_policies.cpp]: Good call!

File: src/libclipper/test/policy_test.cpp (hunk 31 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: Fixed

File: src/management/src/management_frontend.hpp (hunk 35 above)
Comment (yujialuo) [reply src/management/src/management_frontend.hpp]: Fixed

File: src/libclipper/test/selection_policies_test.cpp (hunk 34 above)
Comment (yujialuo) [reply src/libclipper/test/selection_policies_test.cpp]: Fixed

File: src/libclipper/include/clipper/selection_policies.hpp (hunk 22 above)
Comment (yujialuo) [reply src/libclipper/include/clipper/selection_policies.hpp]: Added comments

File: src/libclipper/include/clipper/selection_policies.hpp (hunk 23 above)
Comment (yujialuo) [reply src/libclipper/include/clipper/selection_policies.hpp]: Fixed

File: src/libclipper/src/selection_policies.cpp (hunk 24 above)
Comment (yujialuo) [reply src/libclipper/src/selection_policies.cpp]: Added comments

File: src/libclipper/src/selection_policies.cpp (hunk 25 above)
Comment (yujialuo) [reply src/libclipper/src/selection_policies.cpp]: *UNSOLVED* but what is what are not aiming just at binary classification?

File: src/libclipper/src/selection_policies.cpp (hunk 27 above)
Comment (yujialuo) [reply src/libclipper/src/selection_policies.cpp]: Fixed

File: src/libclipper/test/policy_test.cpp (hunk 29 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: Fixed

File: src/libclipper/test/policy_test.cpp (hunk 30 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: I deleted this because Exp4 selection will always select all models

File: src/libclipper/test/policy_test.cpp (hunk 33 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: Fixed

File: src/libclipper/test/policy_test.cpp (hunk 32 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: *UNSOLVED* I realize something serious is off about my loss function, need help

File: src/libclipper/src/selection_policies.cpp (hunk 26 above)
Comment (dcrankshaw) [reply src/libclipper/src/selection_policies.cpp]: No. Just have it return an empty vector of FeedbackTask.

File: src/libclipper/src/selection_policies.cpp (hunk 25 above)
Comment (dcrankshaw) [reply src/libclipper/src/selection_policies.cpp]: I'm not sure what you're asking? As the code is right now, it won't work for binary classification. It will work for regression where the true average is a useful prediction. We want it to work for at least binary classification.

File: src/libclipper/test/policy_test.cpp [hunk 36]
```diff
@@ -94,8 +98,21 @@ TEST_F(Exp3Test, UpdateTest) {
...
+    auto tasks = Exp3Policy::select_predict_tasks(state, query, 1000);
+    if (model_1.second == tasks.front().model_.second) {
+      select_1 ++;
+    } else if (model_2.second == tasks.front().model_.second) {
+      select_2 ++;
+    } else {
+      select_3 ++;
+    };
+    ASSERT_GE(select_1, select_2);
+    ASSERT_GE(select_2, select_3);
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Does this test pass? I would expect that the asserts should be outside of the for loop?

File: src/libclipper/test/policy_test.cpp [hunk 37]
```diff
@@ -248,7 +286,9 @@ TEST_F(UCBTest, UpdateTest) {
 TEST_F(UCBTest, SelectionTest) {
   auto query = Utility::create_query(models);
   auto tasks = UCBPolicy::select_predict_tasks(state, query, 1000);
-  ASSERT_EQ(model_1.second, tasks.front().model_.second);
+  for (int i=0; i<10; ++i) {
+    ASSERT_EQ(model_1.second, tasks.front().model_.second);
+  }
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: What's going on here? It looks like you're just doing the same assert 10 times without changing anything?

**Review by dcrankshaw — COMMENTED** [review]
I have a couple questions about the unit tests. Other than that it's ready to go.

File: src/libclipper/test/policy_test.cpp (hunk 32 above)
Comment (dcrankshaw) [reply src/libclipper/test/policy_test.cpp]: Okay let's discuss this at the meeting today

File: src/libclipper/test/policy_test.cpp (hunk 37 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: Oops my bad. Fixed.

File: src/libclipper/test/policy_test.cpp (hunk 36 above)
Comment (yujialuo) [reply src/libclipper/test/policy_test.cpp]: Oops my bad. Fixed. (No it didn't pass I need to discuss the implementation of Exp3/Exp4 with you this afternoon)

**Comment by yujialuo** [comment]
Praying for merging.

File: src/libclipper/include/clipper/selection_policies.hpp [hunk 38]
```diff
@@ -8,6 +8,11 @@
 #include "datatypes.hpp"
 #include "task_executor.hpp"
 
+/* IMPORTANT NOTES FOR USING SELECTION POLICIES
```
Comment (dcrankshaw) [inline src/libclipper/include/clipper/selection_policies.hpp]: Make this a doc-comment:
```c++
/**
 * 
 *
 */
```

File: src/libclipper/src/selection_policies.cpp [hunk 39]
```diff
@@ -164,8 +162,8 @@ Exp3Policy::select_feedback_tasks(BanditPolicyState& state, FeedbackQuery feedba
   auto predict_task = PredictTask(feedback.feedback_.input_, selected_model, -1, query_id, -1);
   std::vector<PredictTask> predict_tasks{predict_task};
   // Feedback Task
-  auto feedback_task = FeedbackTask(feedback.feedback_, selected_model, query_id, -1);
-  std::vector<FeedbackTask> feedback_tasks{feedback_task};
+  // auto feedback_task = FeedbackTask(feedback.feedback_, selected_model, query_id, -1);
```
Comment (dcrankshaw) [inline src/libclipper/src/selection_policies.cpp]: delete commented out code

File: src/libclipper/test/policy_test.cpp [hunk 40]
```diff
@@ -11,14 +11,19 @@
 #include <clipper/datatypes.hpp>
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Delete commented out boost/property_tree include

File: src/libclipper/test/policy_test.cpp [hunk 41]
```diff
@@ -55,67 +60,78 @@ class Utility {
...
-    ASSERT_GE(select_1, select_2);
-    ASSERT_GE(select_2, select_3);
   }
-}
+  // Test if times selected model_0 > model_1 > model_2
+  ASSERT_GT(select_0, select_1);
+  ASSERT_GT(select_1, select_2);
 
-TEST_F(Exp3Test, SerializationTest) {
+  /* Serialization Test */
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: Break the serialization test into a separate test case.

File: src/libclipper/test/policy_test.cpp [hunk 42]
```diff
@@ -124,48 +140,46 @@ TEST_F(Exp3Test, SerializationTest) {
...
+  ASSERT_GT(state.model_map_[model_0]["weight"],
+            state.model_map_[model_1]["weight"]);
   ASSERT_GT(state.model_map_[model_1]["weight"],
             state.model_map_[model_2]["weight"]);
-  ASSERT_GT(state.model_map_[model_2]["weight"],
-            state.model_map_[model_3]["weight"]);
-}
 
-TEST_F(Exp4Test, SerializationTest) {
+  /* Serialization Test */
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: separate serialization test into separate test case

File: src/libclipper/test/policy_test.cpp [hunk 43]
```diff
@@ -242,56 +254,53 @@ TEST_F(EpsilonGreedyTest, SerializationTest) {
...
-  for (int i=0; i<10; ++i) {
-    ASSERT_EQ(model_1.second, tasks.front().model_.second);
+  // UCB should always select the optimal bandit
+  for (int i=0; i<select_times_; ++i) {
+    auto tasks = UCBPolicy::select_predict_tasks(state, query, 1000);
+    ASSERT_EQ(model_0.second, tasks.front().model_.second);
   }
-}
 
-TEST_F(UCBTest, SerializationTest) {
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: separate serialization test into separate test case

File: src/libclipper/test/policy_test.cpp [hunk 44]
```diff
@@ -174,66 +188,64 @@ TEST_F(Exp4Test, SerializationTest) {
...
+      select_2 ++;
     };
-    ASSERT_GE(select_1, select_2);
   }
-}
+  // Test if times selected model_0 > model_1
+  ASSERT_GT(select_0, select_1);
 
-TEST_F(EpsilonGreedyTest, SerializationTest) {
+  /* Serialization Test */
```
Comment (dcrankshaw) [inline src/libclipper/test/policy_test.cpp]: separate serialization test into separate test case

File: src/frontends/src/query_frontend.hpp (hunk 2 above)
Comment (dcrankshaw) [reply src/frontends/src/query_frontend.hpp]: Reminder to remove `model_name` and `model_version` from the JSON format comment on this method.

**Review by dcrankshaw — CHANGES_REQUESTED** [review]
You need to fix a couple of the QueryFrontend unit tests. In [`QueryFrontendTest.TestDecodeCorrectUpdate`](https://github.com/yujialuo/Clipper/blob/cd060c3f842faa34c8880140f502a97cb86e9304/src/frontends/src/query_frontend_tests.cpp#L139-L140) remove the model_name and model_version fields (these are no longer part of JSON expected for an update).

In [`QueryFrontendTest.TestDecodeUpdateMissingField `](https://github.com/yujialuo/Clipper/blob/cd060c3f842faa34c8880140f502a97cb86e9304/src/frontends/src/query_frontend_tests.cpp#L151-L152) remove the model_name and label fields. Remove model_name because it's no longer expected. Remove label because this unit test because the label is needed and this test tests that the query_frontend can detect when a required JSON field is missing and throw the correct error.

File: src/frontends/src/query_frontend.hpp (hunk 2 above)
Comment (dcrankshaw) [reply src/frontends/src/query_frontend.hpp]: Specifically, delete [these two lines](https://github.com/yujialuo/Clipper/blob/cd060c3f842faa34c8880140f502a97cb86e9304/src/frontends/src/query_frontend.hpp#L231-L232)

**Review by dcrankshaw — APPROVED** [review]
